sdk/servicebus/service-bus/src/serviceBusMessage.ts

```diff
@@ -983,44 +935,60 @@ export class ServiceBusMessage implements ReceivedMessage {
-        DeadLetterReason: options.deadletterReason,
-        DeadLetterErrorDescription: options.deadLetterErrorDescription
-      };
-    }
+  async deadLetter(propertiesToModify?: DeadLetterOptions & { [key: string]: any }): Promise<void> {
     log.message(
       "[%s] Deadlettering the message with id '%s'.",
       this._context.namespace.connectionId,
       this.messageId
     );
-    if (this._context.requestResponseLockedMessages.has(this.lockToken!)) {
-      await this._context.managementClient!.updateDispositionStatus(
-        this.lockToken!,
-        DispositionStatus.suspended,
-        {
-          deadLetterReason: error.condition,
-          deadLetterDescription: error.description,
-          sessionId: this.sessionId
-        }
-      );
 
-      // Remove the message from the internal map of deferred messages
-      this._context.requestResponseLockedMessages.delete(this.lockToken!);
+    const actualPropertiesToModify: Partial<DeadLetterOptions> = {
+      ...propertiesToModify
+    };
+
+    // these two fields are handled specially and don't need to be in here.
+    delete actualPropertiesToModify.deadLetterErrorDescription;
+    delete actualPropertiesToModify.deadletterReason;
+
+    const dispositionStatusOptions: DispositionStatusOptions = {
+      propertiesToModify: actualPropertiesToModify,
+      deadLetterReason: propertiesToModify?.deadletterReason,
+      deadLetterDescription: propertiesToModify?.deadLetterErrorDescription
+    };
+    return this.settleMessage(DispositionStatus.suspended, dispositionStatusOptions);
+  }
+
+  private async settleMessage(
+    operation: DispositionStatus,
+    options?: DispositionStatusOptions
+  ): Promise<void> {
+    const dispositionType = getDispositionType(operation);
+    const receiver = this.getReceiverFromContext();
+    this.throwIfMessageCannotBeSettled(receiver, dispositionType!);
+
+    const isDeferredMessage = this._context.requestResponseLockedMessages.has(this.lockToken!);
+
+    if (isDeferredMessage || ((!receiver || !receiver.isOpen()) && this.sessionId == undefined)) {
+      await this._context.managementClient!.updateDispositionStatus(this.lockToken!, operation, {
+        ...options,
+        sessionId: this.sessionId
+      });
+      if (isDeferredMessage) {
+        // Remove the message from the internal map of deferred messages
+        this._context.requestResponseLockedMessages.delete(this.lockToken!);
+      }
       return;
     }
-    const receiver = this._context.getReceiver(this.delivery.link.name, this.sessionId);
-    this.throwIfMessageCannotBeSettled(receiver, DispositionType.deadletter);
 
-    return receiver!.settleMessage(this, DispositionType.deadletter, {
-      error: error
-    });
+    return receiver!.settleMessage(this, dispositionType!, options);
+  }
+  private getReceiverFromContext() {
+    if (this.delivery.link) {
+      return this._context.getReceiver(this.delivery.link.name, this.sessionId);
+    } else {
```

Comment:
When do we expect to hit the `else` here? Do we know a case when the delivery object will not have a link?

Reply:
Yup, this method was throwing errors because the link was undefined when settling a deferred message.

Reply:
Previously, `getReceiverFromContext()` was called after the `if-block` for deferred-message-settlement, which is probably why this never threw an error.

Reply:
In that case, I would suggest not having a helper method for this and inline it as well. Because now, the fact that a deferred message cannot have will have no delivery link is tucked away in a layer of redirection. I would recommend the below flow:

```js
const isDeferredMessage = this._context.requestResponseLockedMessages.has(this.lockToken!);
const receiver = isDeferredMessage ? undefined : this._context.getReceiver(this.delivery.link.name, this.sessionId);
this.throwIfMessageCannotBeSettled(receiver, dispositionType!);
```